shared/src/main/scala/scala/xml/factory/XMLLoader.scala

```diff
@@ -25,9 +25,18 @@ trait XMLLoader[T <: Node] {
 
   /* Override this to use a different SAXParser. */
   def parser: SAXParser = {
-    val f = SAXParserFactory.newInstance()
-    f.setNamespaceAware(false)
-    f.newSAXParser()
+    val parser = SAXParserFactory.newInstance()
+
+    parser.setFeature("http://javax.xml.XMLConstants/feature/secure-processing", true)
+    parser.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false)
+    parser.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true)
```

Comment:
I'm not sure how many people use this library to parse HTML, but I'm sure there are some.  Either way having a DOCTYPE in your XML will fail to load:

```scala
    val html =
      """<!DOCTYPE html>
        |<html lang="en">
        |</html>
        |""".stripMargin
    XML.loadString(html)
```

The result is: `org.xml.sax.SAXParseException: DOCTYPE is disallowed when the feature "http://apache.org/xml/features/disallow-doctype-decl" set to true.`

Reply:
It's worth noting that a lowercase `doctype`, which is very common, currently fails already before changing these settings:

```scala
    val html =
      """<!doctype html>
        |<html lang="en">
        |</html>
        |""".stripMargin
    XML.loadString(html)
```

That's because it's malformed, `org.xml.sax.SAXParseException: The markup in the document preceding the root element must be well-formed.`

Reply:
> I'm not sure how many people use this library to parse HTML

I'm comfortable with (not in a 1.1.1 release, but in a 1.2 or a 2.0) requiring those users to explicitly change the default. We might double check if there is suitable documentation in appropriate places.

Reply:
Hmm... that is unfortunate, but it's insecure:
https://www.owasp.org/index.php/XML_External_Entity_(XXE)_Prevention_Cheat_Sheet#General_Guidance

If we leave that enabled we leave ourselves open to DOS attacks. The cleanest solution would probably be to document this and have people that need it provide their own parser instance. You just have to do XML.withSAXParser(myParser).load(html) instead, so it shouldn't be a huge burden.

Reply:
the comment (though it/s not a Scaladoc comment) says `/* Override this to use a different SAXParser. */`, is that not the right recommendation to make...?

Reply:
also if the recommendation is `XML.withSAXParser(myParser)` is there a way to obtain a `myParser` that begins with the secure defaults? is a refactor needed here to make that available?

Reply:
> is that not the right recommendation to make...?

I'm having the same thoughts along those lines... comment coming...

Reply:
> also if the recommendation is XML.withSAXParser(myParser) is there a way to obtain a myParser that begins with the secure defaults? is a refactor needed here to make that available?

You could add a "safeParser" factory method somewhere that supplies one, but that would mean the user would have to explicitly use that. It's still unsafe by default and basically the current situation where the user has to be aware of this issue and work around it. I'm not so sure that's such a great idea.

Reply:
> It's still unsafe by default

we definitely want to make things safe by default, absolutely.

the question is rather how best to support people who want to use the safe settings as a starting point but then tweak them.

Reply:
>we definitely want to make things safe by default, absolutely.

Ah, sorry. I misunderstood you.

So the idea is some variation of:
```scala
def parser(f: SAXParserFactory => SAXParserFactory = identity)
```
... where the factory has safe defaults already applied to allow for user customization?
That makes a lot of sense.

Reply:
Yeah, something like that would be awesome.

Reply:
There's a tradeoff of breaking runtime behavior and securing by default here. It is a security issue, but we should also be mindful that not everyone who deals with XML deals with XML documents of untrusted origin. So this is more nuanced case-by-case tradeoff.

Given that this issue has been around, and if informed users have been disabling external DTD loading features manually, then the rest are casual/no-so-informed users by process of elimination. I am not too sure if they will read the release notes whether it comes out in version 1.1.x or 1.2.x.

A potential way of dealing with this is to use static typing. As in we would deprecate any methods that are current unsafe during 1.1.x and provide safer one as the alternative. In 1.2.x you can remove the current unsafe methods.

See also https://github.com/scala/scala-xml/issues/17 where this was discussed.

Reply:
Thanks for weighing in on this, Eugene.

I put a general comment on this discussion at the top-level (outside of the patch review), see below https://github.com/scala/scala-xml/pull/177#issuecomment-367779425

Reply:
only update on this? This is a pretty significant security flaw and libraries like Jackson have already addressed it. I don't see any traction on this since February and a security flaw like this honestly should be treated as a hotfix not as a lazy trickle down enhancement request!